opengrok-web/src/main/java/org/opengrok/web/api/v1/controller/FileController.java

```diff
@@ -0,0 +1,169 @@
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at LICENSE.txt.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+/*
+ * Copyright (c) 2020 Oracle and/or its affiliates. All rights reserved.
+ */
+
+package org.opengrok.web.api.v1.controller;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import org.apache.lucene.document.Document;
+import org.apache.lucene.queryparser.classic.ParseException;
+import org.opengrok.indexer.analysis.AbstractAnalyzer;
+import org.opengrok.indexer.configuration.RuntimeEnvironment;
+import org.opengrok.indexer.search.QueryBuilder;
+import org.opengrok.web.api.v1.filter.CorsEnable;
+import org.opengrok.web.api.v1.filter.PathAuthorized;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.opengrok.indexer.index.IndexDatabase.getDocument;
+
+@Path(FileController.PATH)
+public class FileController {
+
+    public static final String PATH = "/file";
+
+    private static final RuntimeEnvironment env = RuntimeEnvironment.getInstance();
+
+    private ArrayList<String> lines = new ArrayList<>();
+
+    static class LineDTO {
+        @JsonProperty
+        private String line;
+        @JsonProperty
+        private int number;
+
+        // for testing
+        LineDTO() {
+        }
+
+        LineDTO(String line, int num) {
+            this.line = line;
+            this.number = num;
+        }
+
+        public String getLine() {
+            return this.line;
+        }
+
+        public int getNumber() {
+            return this.number;
+        }
+    }
+
+    private static File getFile(String path, HttpServletResponse response) throws IOException {
+        if (path == null) {
+            if (response != null) {
+                response.sendError(HttpServletResponse.SC_NOT_ACCEPTABLE, "Missing path parameter");
+            }
+            return null;
+        }
+
+        File file = new File(env.getSourceRootFile(), path);
+        if (!file.isFile()) {
+            if (response != null) {
+                response.sendError(HttpServletResponse.SC_NOT_FOUND, "File not found");
+            }
+            return null;
+        }
+
+        return file;
+    }
+
+    @GET
+    @CorsEnable
+    @PathAuthorized
+    @Path("/content")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Object getContent(@Context HttpServletRequest request,
```

Comment:
Not sure the utility of a web method that just splits a file using standard Java `readline()` and returns it encoded in JSON. Why not just return the raw content more efficiently?

Reply:
I was wondering about this myself. This was originally suggested in #3042 by @jbaek7023 , perhaps there is some value in that for him.

Reply:
I think @tulinkry's suggestion to use accept headers makes sense.

Reply:
If it’s returning a split meant to align with OpenGrok’s analysis, then it should be using `SourceSpllitter` with any necessary project-configured `ExpandTabsReader` via a `ReaderWrapper`. (There is a TODO to comprehensively accommodate the various line/paragraph/page white space separators. E.g. see Erlang.)

Otherwise I recommend to defer the JSON format and let the client handle the raw source text on its own.

In this patch it’s only `readline()`, and that is not what OpenGrok uses. Also `readline()` does not preserve end of line space, so the sum of the lines will not equal the original document, and the client will have no way to know original line endings. (With some languages like Erlang they could be varied in the original.)

Reply:
True, JSON removed.

Reply:
> and the client will have no way to know original line endings

Not true if implemented with content negotiation for both raw format and json format. When the client cares, he can choose the raw format. When he does not, he just likes to have a list of lines, he chooses json with split lines.